File: php/eddington.php
#!/usr/bin/env php
<?php
function eddington($rides) {
  
  $n = count($rides);
  
  if (!$n) return 0;

  rsort($rides);
  
  $E = 0;
  for (; $E < $n; $E++) {
    if ($rides[$E] < $E + 1) break;
  }

  return $E;
}

if ($argc > 1) {

  $lines = file($argv[1]);
  foreach ($lines as $line) {
    $rides[] = (float) $line;
  }

  echo eddington($rides)."\n";

}

exit;

?>

File: lua/eddington.lua
#!/usr/bin/env lua
function eddington (file)
    local e, above = 0, 0
    local dist_table = {}

    for line in io.lines(file) do
        local ride = math.floor(line)

        if ride > e then
            above = above + 1
            dist_table[ride] = (dist_table[ride] or 0) + 1
        end

        if above > e then
            e = e + 1
            if dist_table[e] ~= nil then
                above = above - dist_table[e]
                dist_table[e] = nil
            end
        end

    end

    return e
end

function main ()
    local file = arg[1]
    print(eddington(file))
end

if not package.loaded['eddington'] then
    main()
end
